### DISCUSSION TITLE
Generate uuid columns

### DISCUSSION THREAD

**Discussion by kerruba-milkman** [saulpw/visidata#1564] [Q&A] [answered]
Hello, 

I'm a newbie of visidata - and in general not a very expert Python developer - but I'm trying to define a command in visidata to generate 
a uuid column.

I was able to generate uuids using this function in my `.visidatarc`
```
import uuid


#TODO Need to generate value only once and not every time the sheet is updated
def uuidColumn(sourceCol):
    return Column(sourceCol.name+"_uuid", type=str, getter=lambda _, r: uuid.uuid4() )

Sheet.addCommand(None, 'add-uuid-column', 'addColumn(uuidColumn(cursorCol), cursorColIndex+1)', 'generate a uuid v4 column')
```

the problem is though that I don't want the cell to be recalculated everytime I move my cursor on the sheet. I've also tested the `cache=True` option, and even if it seems to be quite a good candidate to solve the uuid re-generation, there are actions that trigger the generation anyway. 
Instead what I want to achieve is generate the uuid once and that should remain, no re-calculation done

Any help would be fantastic :-)

**Comment by saulpw** [comment, marked as answer]
Hey @kerruba-milkman, that's an interesting question.  One thing you could do is "freeze" (with `'`) the column after creating it, and then hide the original column and use the new frozen column for everything.  You could also incorporate this into your command by using `addColumnAtCursor(StaticColumn(uuidColumn(cursorCol)))` as the execstr.

Alternatively, you could follow the `setcol-` command pattern rather than the `addcol-` pattern, since you want them to be set once and never recalculated.  For instance:

```
@Column.api
@asyncthread
def setcol_uuid(col, rows):
    col.setValues(rows, *[uuid.uuid4() for r in rows])

Sheet.addCommand('', 'setcol-uuid', 'cursorCol.setcol_uuid(someSelectedRows)')
```

(The `@asyncthread` is needed in case the number of selected rows is large.)

Now you can create a blank column with `za` and use `setcol-uuid`.

**Reply by kerruba-milkman** [reply]
Hey @saulpw, thanks a lot for your answer, it was definitely useful, and I was able to perform the task with your suggestions.
I still have a couple of questions/notes:
1. The first snippet isn't really working, as I'm getting all the rows in the new column reporting an error `'NoneType' has no attribute 'defer'`. Any suggestion on this?
2. The second snippet worked fine, only had to change the method parameter from `col` to `self`.

Thanks again for all the help provided.

BTW, the search box on the API docs isn't really working the way I was expecting, or isn't working at all. I was searching for a reference to the `cursorCol`, and it couldn't find any reference.

**Reply by lkhrs** [reply]
Hi @kerruba-milkman, a fix is live now for the search issue on the API docs. How does it look on your end? https://www.visidata.org/docs/api/search.html?q=cursorcol

**Reply by kerruba-milkman** [reply]
Hi @lkhrs, now seems to work fine thanks

**Reply by kerruba-milkman** [reply]
> 1. The first snippet isn't really working, as I'm getting all the rows in the new column reporting an error `'NoneType' has no attribute 'defer'`. Any suggestion on this?

Actually I found what was the issue.
Basically, I have to pass the sheet object to the `Column`  kwargs in order to be able to refer it later in the `self.sheet.defer` which was giving me the error.

The final command is therefore this

```
import uuid

def uuidColumn(sourceCol):
    return Column(sourceCol.name+"_uuid", type=str, getter=lambda c, r: uuid.uuid4(), sheet=sourceCol.sheet)

Sheet.addCommand(None, 'add-uuid-column', 'sheet.addColumnAtCursor(StaticColumn(uuidColumn(cursorCol)))', 'generate a uuid v4 column')
```

**Reply by saulpw** [reply]
Nice catch, @kerruba-milkman!  yeah, `addColumn` sets that, which is why it's necessary to call it instead of appending to `columns` directly.  And apparently you can only freeze columns that have their `sheet` set, which makes sense.  Thanks for digging into it!
